### ISSUE TITLE
`no-unused-expressions` has no way to warn on `void foo`

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#9845]
**What version are you using?**
latest

**What did you do?**
enabled `no-unused-expressions`, and expected `void String;` to be an error.

**What happened?**
It was not an error.

**What did you expect to happen?**
An error.

Looking at the exceptions in [the docs](https://eslint.org/docs/rules/no-unused-expressions), `new C` makes sense because constructors can have side effects (as could any function invocation or property access); `delete a.b` makes sense because `delete` has an effect; assignment makes sense. However, while some people might want `void a` to be used, that statement has no actual effect, and I'd like to be able to warn against it.

Could we add an option?

**Comment by not-an-aardvark** [comment]
This seems like a bug, right? If I'm understanding correctly, `void something` as an `ExpressionStatement` would always be unused, because it could just be replaced with `something`.

**Comment by ljharb** [comment]
Yes, that seems right to me. However, its explicit inclusion in the docs suggests that it was intentional.

Either a fix or an option is fine by me :-)

**Comment by not-an-aardvark** [comment]
You're right, it seems like this was intentional -- it was introduced in https://github.com/eslint/eslint/commit/928b6aeb1d499628212b44b486cca3557e53bc35 as a fix for https://github.com/eslint/eslint/issues/431.

I don't really agree with the reasoning in that issue -- it seems like `void` is being used to work around the [`no-new`](https://eslint.org/docs/rules/no-new) rule, but a disable comment would probably be a better way to do that because the use of `void` doesn't really make sense here. However, if we do fix this by default it would be a breaking change.

**Comment by platinumazure** [comment]
I would be :+1: for an option, myself.

**Comment by ljharb** [comment]
The original intention seemed to be to emulate the jshint behavior of "void means you didn't make a mistake" - however, eslint has a much more explicit means of conveying that: override comments.

**Comment by taion** [comment]
BTW, this came up in the context of my wanting to find a way to mark variables as intentionally unused to avoid having to enable `ignoreRestSiblings` for no-unused. My intended pattern was:

```js
const { foo, bar, ...rest } = whatever;
void bar;
```

But actually this is an ugly pattern and I'm not going to use it.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.

**Comment by taion** [comment]
I'm inclined to agree with @ljharb in https://github.com/eslint/eslint/issues/9845#issuecomment-357528686. If the `void` handling here is only for the sake of a pragma, then we should remove it, as there are better ways to do pragmas now.
